src/parcsr_ls/par_mod_multi_interp.c

```diff
@@ -0,0 +1,1305 @@
+   }
+   else
+   {
+      total_global_cpts = num_cpts_global[1];
+   }
+   /* Generate pass marker array */
+
+   pass_marker = hypre_CTAlloc(HYPRE_Int, n_fine, HYPRE_MEMORY_HOST);
+   /* contains pass numbers for each variable according to original order */
+   pass_order = hypre_CTAlloc(HYPRE_Int, n_fine, HYPRE_MEMORY_HOST);
+   /* contains row numbers according to new order, pass 1 followed by pass 2 etc */
+   fine_to_coarse = hypre_CTAlloc(HYPRE_Int, n_fine, HYPRE_MEMORY_HOST);
+   /* reverse of pass_order, keeps track where original numbers go */
+   points_left = hypre_CTAlloc(HYPRE_Int, n_fine, HYPRE_MEMORY_HOST);
+   /* contains row numbers of remaining points, auxiliary */
+   pass_starts = hypre_CTAlloc(HYPRE_Int, 10, HYPRE_MEMORY_HOST);
+   /* contains beginning for each pass in pass_order field, assume no more than 10 passes */
+
+   P_diag_i = hypre_CTAlloc(HYPRE_Int, n_fine+1, HYPRE_MEMORY_HOST);
+   P_offd_i = hypre_CTAlloc(HYPRE_Int, n_fine+1, HYPRE_MEMORY_HOST);
+
+   cnt = 0;
+   remaining = 0;
+   for (i=0; i < n_fine; i++)
+   {
+      if (CF_marker[i] == 1)
+      {
+         pass_marker[i] = 1;
+         P_diag_i[i+1] = 1;
+         P_offd_i[i+1] = 0;
+         fine_to_coarse[i] = cnt;
+         pass_order[cnt++] = i;
+      }
+      else
+      {
+         points_left[remaining++] = i;
+      }
+   }
+   pass_starts[0] = 0;
+   pass_starts[1] = cnt;
+
+   if (num_functions > 1)
+   {
+      dof_func_offd = hypre_CTAlloc(HYPRE_Int,  num_cols_offd_A, HYPRE_MEMORY_HOST);
+      index = 0;
+      num_sends = hypre_ParCSRCommPkgNumSends(comm_pkg);
+      int_buf_data = hypre_CTAlloc(HYPRE_Int,  hypre_ParCSRCommPkgSendMapStart(comm_pkg, num_sends), HYPRE_MEMORY_HOST);
+      for (i = 0; i < num_sends; i++)
+      {
+         startc = hypre_ParCSRCommPkgSendMapStart(comm_pkg, i);
+         for (j = startc; j < hypre_ParCSRCommPkgSendMapStart(comm_pkg, i+1); j++)
+         {
+            int_buf_data[index++] = dof_func[hypre_ParCSRCommPkgSendMapElmt(comm_pkg,j)];
+         }
+      }
+
+      comm_handle = hypre_ParCSRCommHandleCreate( 11, comm_pkg, int_buf_data, dof_func_offd);
+
+      hypre_ParCSRCommHandleDestroy(comm_handle);
+
+      hypre_TFree(int_buf_data, HYPRE_MEMORY_HOST);
+   }
+
+   if (num_cols_offd_A)
+   {
+      pass_marker_offd = hypre_CTAlloc(HYPRE_Int,  num_cols_offd_A, HYPRE_MEMORY_HOST);
+      index = 0;
+      num_sends = hypre_ParCSRCommPkgNumSends(comm_pkg);
+      int_buf_data = hypre_CTAlloc(HYPRE_Int,  hypre_ParCSRCommPkgSendMapStart(comm_pkg, num_sends), HYPRE_MEMORY_HOST);
+      for (i = 0; i < num_sends; i++)
+      {
+         startc = hypre_ParCSRCommPkgSendMapStart(comm_pkg, i);
+         for (j = startc; j < hypre_ParCSRCommPkgSendMapStart(comm_pkg, i+1); j++)
+         {
+            int_buf_data[index++] = pass_marker[hypre_ParCSRCommPkgSendMapElmt(comm_pkg,j)];
+         }
+      }
+
+      comm_handle = hypre_ParCSRCommHandleCreate( 11, comm_pkg, int_buf_data, pass_marker_offd);
```

Comment:
@ulrikeyang This looks strange: `hypre_ParCSRCommHandleCreate` inside the if-statement. A process can have zero `num_cols_offd_A` but it may still need to participate in the communication. Am I right on this?

Reply:
You are right. This should be changed to 
if (num_procs > 1)
to be safe.
Thanks for finding this.

Reply:
Will pass_marker_offd set to NULL if num_cols_offd_A is 0? Or do we need to add extra precautions for this special case of num_Cols_offd_A = 0 and still needing to send data.be sent?

Reply:
@ulrikeyang If you were talking about this line
```
      pass_marker_offd = hypre_CTAlloc(HYPRE_Int, num_cols_offd_A, HYPRE_MEMORY_HOST);
```
yes, `pass_marker_offd` is `NULL` when `num_cols_offd_A==0`. I think after the change it will be fine even when `num_cols_offd_A==0`. More thoughts? Thanks!